crates/notion-core/src/distro/package.rs

```diff
@@ -0,0 +1,403 @@
+                &self.image_dir.clone().into_os_string(),
+                &image.path()?,
+            )
+        };
+
+        let output = install_cmd.output().with_context(install_error)?;
+        if !output.status.success() {
+            throw!(ErrorDetails::PackageInstallFailed {
+                cmd: format!("{:?}", install_cmd),
+                status: output.status
+            });
+        }
+
+        self.write_config_and_shims(&platform)?;
+
+        Ok(())
+    }
+
+    fn package_config(&self, platform_spec: &PlatformSpec) -> PackageConfig {
+        PackageConfig {
+            name: self.name.to_string(),
+            version: self.version.clone(),
+            platform: platform_spec.clone(),
+            bins: self
+                .bins
+                .iter()
+                .map(|(name, _path)| name.to_string())
+                .collect(),
+        }
+    }
+
+    fn bin_config(
+        &self,
+        bin_name: String,
+        bin_path: String,
+        platform_spec: &PlatformSpec,
+    ) -> BinConfig {
+        BinConfig {
+            name: bin_name,
+            package: self.name.to_string(),
+            version: self.version.clone(),
+            path: bin_path,
+            platform: platform_spec.clone(),
+        }
+    }
+
+    fn write_config_and_shims(&self, platform_spec: &PlatformSpec) -> Fallible<()> {
+        self.package_config(&platform_spec).to_serial().write()?;
+        for (bin_name, bin_path) in self.bins.iter() {
+            self.bin_config(bin_name.to_string(), bin_path.to_string(), &platform_spec)
+                .to_serial()
+                .write()?;
+            // create a link to the shim executable
+            shim::create(&bin_name)?;
+        }
+        Ok(())
+    }
+}
+
+impl Installer {
+    pub fn cmd(&self) -> Command {
+        match self {
+            Installer::Npm => {
+                let mut command = Command::new("npm");
+                command.args(&["install", "--only=production"]);
+                command
+            }
+            Installer::Yarn => {
+                let mut command = Command::new("yarn");
+                command.args(&["install", "--production"]);
+                command
+            }
+        }
+    }
+}
+
+/// Information about a user tool.
+pub struct UserTool {
+    pub bin_path: PathBuf,
+    pub image: Image,
+}
+
+impl UserTool {
+    pub fn from_config(bin_config: BinConfig, session: &mut Session) -> Fallible<Self> {
+        let image_dir =
+            path::package_image_dir(&bin_config.package, &bin_config.version.to_string())?;
+        // canonicalize because path is relative, and sometimes uses '.' char
+        let bin_path = image_dir.join(bin_config.path).canonicalize().unknown()?;
+
+        Ok(UserTool {
+            bin_path,
+            image: bin_config.platform.checkout(session)?,
+        })
+    }
+}
+
+pub fn user_tool(tool_name: &str, session: &mut Session) -> Fallible<Option<UserTool>> {
```

Comment:
Stylistically I would probably make this a "static method" of `UserTool`, like `UserTool::new()` or `UserTool::from_name()` or something like that. Your call though.

Reply:
That's a good idea, I'll to `UserTool::from_name()` since there is already `UserTool::from_config()`